label_studio/io_storages/aperturedb/models.py

```diff
@@ -0,0 +1,557 @@
+    _db = None
+
+    secure_fields = ["password", "token"]
+
+    def _response_status(self, response):
+        if isinstance(response, list):
+            return max([self._response_status(val) for val in response])
+        elif isinstance(response, dict):
+            if "status" in response:
+                return response["status"]
+            for val in response.values():
+                return self._response_status(val)
+
+    def get_connection(self):
+        if self._db is None:
+            with self._db_lock:
+                if self._db is None:
+                    self._db = Connector.Connector(
+                        str(self.hostname),
+                        self.port,
+                        user=str(self.username),
+                        password=str(self.password) if self.password else "",
+                        token=str(self.token) if self.token else "",
+                        use_ssl=self.use_ssl,
+                    )
+        return self._db
+
+    def validate_connection(self, client=None):
+        db = self.get_connection()
+        res, _ = db.query([{"GetStatus": {}}])
+        if self._response_status(res) != 0:
+            raise ValueError(
+                f"Failed to connect to ApertureDB: {db.get_last_response_str()}")
+
+    class Meta:
+        abstract = True
+
+
+class ApertureDBImportStorageBase(ApertureDBStorageMixin, ImportStorage):
+    url_scheme = "none"
+
+    constraints = models.TextField(
+        _("constraints"),
+        blank=True,
+        null=True,
+        help_text="ApertureDB FindImage constraints (see https://docs.aperturedata.io/query_language/Reference/shared_command_parameters/constraints)",
+    )
+
+    predictions = models.BooleanField(
+        _("predictions"), default=False, help_text="Load bounding box predictions from ApertureDB?"
+    )
+
+    pred_constraints = models.TextField(
+        _("constraints"),
+        blank=True,
+        null=True,
+        help_text="ApertureDB constraints on bounding box predictions (see https://docs.aperturedata.io/query_language/Reference/shared_command_parameters/constraints)",
+    )
+
+    limit = models.PositiveIntegerField(
+        _("limit"), null=True, blank=True, help_text="Maximum number of tasks", default=DEFAULT_LIMIT)
+
+    as_format_jpg = models.BooleanField(
+        _("as_format_jpg"),
+        default=False,
+        help_text="Convert images to JPEG format when loading from ApertureDB",
+    )
+
+    def iterkeys(self):
+        db = self.get_connection()
+
+        batch = 100
+
+        offset = 0
+        find_images = {
+            "uniqueids": True,
+            "blobs": False,
+            "limit": batch,
+            "constraints": {"width": [">", 0], "height": [">", 0]}
+        }
+
+        # Concatenate user constraints with our internal width/height requirement
+        if self.constraints:
+            find_images["constraints"].update(
+                json.loads(str(self.constraints)))
+
```

Comment:
I assume at this level of development if someone botches up the constraints we don't have a nice way of telling them

Reply:
The UI enforces valid JSON, but yes this could easily result in a malformed query.